src/Analyzer.Reports/ReportsHelper.cs

```diff
@@ -0,0 +1,121 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Azure.Templates.Analyzer.Types;
+using Microsoft.CodeAnalysis.Sarif;
+
+namespace Microsoft.Azure.Templates.Analyzer.Reports
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    public class ReportsHelper
+    {
+        /// <summary>
+        /// Get all distinct results in an evaluation organized by source location file
+        /// </summary>
+        /// <param name="evaluations">Evaluations to get results for</param>
+        /// <param name="filesToSkip">Files to not include results from</param>
+        /// <returns></returns>
+        public static Dictionary<string, List<(IEvaluation, IList<IResult>)>> GetResultsByFile(
+            IEnumerable<Types.IEvaluation> evaluations,
+            IEnumerable<string> filesToSkip)
+        {
+            return GetResultsByFile(evaluations, filesToSkip, out int _);
+        }
+
+        /// <summary>
+        /// Get all distinct results in an evaluation organized by source location file
+        /// </summary>
+        /// <param name="evaluations">Evaluations to get results for</param>
+        /// <param name="filesToSkip">Files to not include results from</param>
+        /// <param name="passedEvaluations">Out parameter that gives number of passed evaluations</param>
+        /// <returns></returns>
+        public static Dictionary<string, List<(IEvaluation, IList<IResult>)>> GetResultsByFile(
+            IEnumerable<Types.IEvaluation> evaluations,
+            IEnumerable<string> filesToSkip,
+            out int passedEvaluations)
+        {
+            var resultsByFile = new Dictionary<string, List<(IEvaluation, IList<IResult>)>>();
+            passedEvaluations = 0;
+
+            foreach (var evaluation in evaluations)
+            {
+                if (!evaluation.Passed)
+                {
+                    (var actualFile, var failedResults) = GetResultsByFileInternal(evaluation);
+
+                    // a file's results may have already been output if analyzing directory
+                    if (filesToSkip.Contains(actualFile))
+                    {
+                        continue;
+                    }
+
+                    if (!resultsByFile.ContainsKey(actualFile))
+                    {
+                        resultsByFile[actualFile] = new List<(IEvaluation, IList<IResult>)>();
```

Comment:
In order to do this I would probably implement a generic expression type to contain the IEnumerable<IExpression> that contains the unique results (and refactor the other methods to keep the outer expression of the IResults). Other alternatives creating instances of the existing evaluation types would be more involved I think.

So this would add some additional complexity/code I don't think really improves the readability/simplicity of the code overall since if requires a good amount of additional code/refactoring. Lmk if you want to discuss further.